controlplane/eks/controllers/awsmanagedcontrolplane_controller.go

```diff
@@ -455,3 +463,101 @@ func (r *AWSManagedControlPlaneReconciler) managedClusterToManagedControlPlane(c
 		}
 	}
 }
+
+func (r *AWSManagedControlPlaneReconciler) getRolesForWorkers(ctx context.Context, managedScope *scope.ManagedControlPlaneScope) (map[string]struct{}, error) {
+	// previously this was the default role always added to the IAM authenticator config
+	// we'll keep this to not break existing behavior for users
+	allRoles := map[string]struct{}{
+		fmt.Sprintf("nodes%s", iamv1.DefaultNameSuffix): {},
+	}
+	if err := r.getRolesForMachineDeployments(ctx, managedScope, allRoles); err != nil {
+		return nil, fmt.Errorf("failed to get roles from machine deployments %w", err)
+	}
+	if err := r.getRolesForMachinePools(ctx, managedScope, allRoles); err != nil {
+		return nil, fmt.Errorf("failed to get roles from machine pools %w", err)
+	}
+	return allRoles, nil
+}
+
+func (r *AWSManagedControlPlaneReconciler) getRolesForMachineDeployments(ctx context.Context, managedScope *scope.ManagedControlPlaneScope, allRoles map[string]struct{}) error {
+	deploymentList := &clusterv1.MachineDeploymentList{}
+	selectors := []client.ListOption{
+		client.InNamespace(managedScope.Namespace()),
+		client.MatchingLabels{
+			clusterv1.ClusterLabelName: managedScope.Cluster.Name,
+		},
+	}
+	err := r.Client.List(ctx, deploymentList, selectors...)
+	if err != nil {
+		return fmt.Errorf("failed to list machine deployments for cluster %s/%s: %w", managedScope.Namespace(), managedScope.Cluster.Name, err)
+	}
+
+	for _, deployment := range deploymentList.Items {
+		ref := deployment.Spec.Template.Spec.InfrastructureRef
+		if ref.Kind != "AWSMachineTemplate" {
+			continue
+		}
+		awsMachineTemplate := &infrav1.AWSMachineTemplate{}
+		err := r.Client.Get(ctx, client.ObjectKey{
+			Name:      ref.Name,
+			Namespace: managedScope.Namespace(),
+		}, awsMachineTemplate)
+		if err != nil {
+			return fmt.Errorf("failed to get AWSMachine %s/%s: %w", ref.Namespace, ref.Name, err)
+		}
+		instanceProfile := awsMachineTemplate.Spec.Template.Spec.IAMInstanceProfile
+		if _, ok := allRoles[instanceProfile]; !ok && instanceProfile != "" {
+			allRoles[instanceProfile] = struct{}{}
+		}
+	}
+	return nil
+}
+
+func (r *AWSManagedControlPlaneReconciler) getRolesForMachinePools(ctx context.Context, managedScope *scope.ManagedControlPlaneScope, allRoles map[string]struct{}) error {
+	machinePoolList := &expclusterv1.MachinePoolList{}
+	selectors := []client.ListOption{
+		client.InNamespace(managedScope.Namespace()),
+		client.MatchingLabels{
+			clusterv1.ClusterLabelName: managedScope.Cluster.Name,
+		},
+	}
+	err := r.Client.List(ctx, machinePoolList, selectors...)
+	if err != nil {
+		return fmt.Errorf("failed to list machine pools for cluster %s/%s: %w", managedScope.Namespace(), managedScope.Cluster.Name, err)
+	}
+	for _, pool := range machinePoolList.Items {
+		ref := pool.Spec.Template.Spec.InfrastructureRef
+		switch ref.Kind {
+		case "AWSMachinePool":
```

Comment:
nit: Can we refactor both the cases to a separate func?

Reply:
```golang
		switch ref.Kind {
		case "AWSMachinePool":
                    err := getRolesForAWSMachinePool(ctx, allRoles)
                    
                 case "AWSManagedMachinePools":
                   err := getRolesForAWSManagedMachinePool(ctx, allRoles)
 ```
 
 did you mean something like this?

Reply:
Yes

Reply:
@faiq i think this comment is still applicable.

Reply:
refactored